homeassistant/components/xiaomi_miio/sensor.py

```diff
@@ -637,23 +648,36 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
 class XiaomiGenericSensor(XiaomiCoordinatedMiioEntity, SensorEntity):
     """Representation of a Xiaomi generic sensor."""
 
-    def __init__(
-        self,
-        name,
-        device,
-        entry,
-        unique_id,
-        coordinator,
-        description: XiaomiMiioSensorDescription,
-    ):
+    entity_description: XiaomiMiioSensorDescription
+
+    def __init__(self, name, device, entry, unique_id, coordinator, description):
         """Initialize the entity."""
         super().__init__(name, device, entry, unique_id, coordinator)
+        self.entity_description = description
         self._attr_unique_id = unique_id
-        self.entity_description: XiaomiMiioSensorDescription = description
+        self._attr_native_value = self._determine_native_value()
 
     @property
-    def native_value(self):
-        """Return the state of the device."""
+    def extra_state_attributes(self):
+        """Return the state attributes."""
+        return {
+            attr: self._extract_value_from_attribute(self.coordinator.data, attr)
+            for attr in self.entity_description.attributes
+            if hasattr(self.coordinator.data, attr)
```

Comment:
We should add a check if the extracted value is None here so we filter out those items in the state attributes. We can use a walrus operator.

Reply:
```
In [1]: attributes = ["one", "two", "three"]

In [2]: data = {"one": 1, "two": 2, "three": None}

In [3]: {
   ...:     attr: value
   ...:     for attr in attributes
   ...:     if (value := data.get(attr)) is not None
   ...: }
Out[3]: {'one': 1, 'two': 2}
```

Reply:
Would it be possible to extract the value in a callback instead of in the property. We try to avoid doing side effects like logging in entity state properties.